src/rgw/rgw_bucket.cc

```diff
@@ -2510,13 +2510,28 @@ int RGWBucketInstanceMetadataHandler::do_put(RGWSI_MetaBackend_Handler::Op *op,
                                              RGWMetadataObject *obj,
                                              RGWObjVersionTracker& objv_tracker,
                                              optional_yield y,
-                                             RGWMDLogSyncType type)
+                                             RGWMDLogSyncType type, bool from_remote_zone)
 {
   RGWMetadataHandlerPut_BucketInstance put_op(svc.bucket->ctx(), this, op, entry, obj,
-                                              objv_tracker, y, type);
+                                              objv_tracker, y, type, from_remote_zone);
   return do_put_operate(&put_op);
 }
 
+void init_default_bucket_layout(CephContext *cct, RGWBucketInfo& info, const RGWZone& zone,
+                                const uint32_t* pmaster_num_shards) {
+  info.layout.current_index.gen = 0;
+  info.layout.current_index.layout.normal.hash_type = rgw::BucketHashType::Mod;
+  info.layout.current_index.layout.type = rgw::BucketIndexType::Normal;
+
+  if (pmaster_num_shards != nullptr) {
+    info.layout.current_index.layout.normal.num_shards = *pmaster_num_shards;
+  } else {
```

Comment:
lets drop this `pmaster_num_shards` argument, and just have `init_default_bucket_layout()` set its default shard count. then after `RGWRados::create_bucket()` calls `init_default_bucket_layout()`, it can overwrite that shard count if it has a `pmaster_num_shards`

Reply:
@cbodley should we just do away with `pmaster_num_shards` altogether? From what I can see, it is used only when the create_bucket() call is made from non-master zone. It uses `pmaster_num_shards` to get the value of master's num_shards. Now that we want to keep the layouts local to each zone, we might not have a use for this variable.

Reply:
@cbodley I went ahead and removed `pmaster_num_shards` anyway. pushed with the rest of the changes requested. Could you please review again?